File: calendario/static/js/consentimiento.js
/* Consentimiento de cookies propio, en sustitución de Cookiebot.
 *
 * Los valores por defecto de Consent Mode NO se ponen aquí: van en un script
 * síncrono en el <head> (_includes/_consentimiento.html), porque tienen que
 * estar antes de que cargue cualquier etiqueta de Google. Este fichero solo se
 * ocupa de la interfaz y de comunicar la decisión.
 *
 * Al decidir se hacen tres cosas, en este orden:
 *   1. `gtag('consent','update', …)` con las cuatro categorías traducidas a las
 *      claves de Consent Mode v2.
 *   2. Un `cookie_consent_<categoría>` al dataLayer por cada categoría
 *      aceptada, y un `cookie_consent_update` al final. Son exactamente los
 *      eventos que empuja la plantilla de Cookiebot en GTM, así que los
 *      triggers que ya existen en los contenedores siguen funcionando sin
 *      tocarlos.
 *   3. Se guarda la decisión en una cookie propia con su fecha y su versión.
 */
(function (w, d) {
  var COOKIE = 'cqx_consent';
  var cfg = w.__CONSENTIMIENTO__ || {};
  var caja = d.getElementById('cqx-consent');
  if (!caja) return;

  function gtag() { w.dataLayer = w.dataLayer || []; w.dataLayer.push(arguments); }

  // ---------------------------------------------------------------- guardado

  function leer() {
    var m = d.cookie.match('(^|;)\\s*' + COOKIE + '\\s*=\\s*([^;]+)');
    if (!m) return null;
    try {
      var v = JSON.parse(decodeURIComponent(m[2]));
      // Si las categorías o los textos han cambiado, lo que se aceptó ya no es
      // lo que hay, así que se vuelve a preguntar.
      return v && v.v === cfg.version ? v : null;
    } catch (e) { return null; }
  }

  function guardar(c) {
    var valor = encodeURIComponent(JSON.stringify({
      v: cfg.version, p: c.preferences ? 1 : 0, s: c.statistics ? 1 : 0,
      m: c.marketing ? 1 : 0, t: new Date().toISOString(),
    }));
    // Doce meses, que es lo que recomienda la AEPD y lo que hacía Cookiebot.
    var caduca = new Date(Date.now() + 365 * 24 * 60 * 60 * 1000).toUTCString();
    var seguro = location.protocol === 'https:' ? ';Secure' : '';
    d.cookie = COOKIE + '=' + valor + ';path=/;expires=' + caduca + ';SameSite=Lax' + seguro;
  }

  // ---------------------------------------------------------------- difusión

  function comunicar(c) {
    var conceder = function (b) { return b ? 'granted' : 'denied'; };
    gtag('consent', 'update', {
      ad_storage: conceder(c.marketing),
      ad_user_data: conceder(c.marketing),
      ad_personalization: conceder(c.marketing),
      analytics_storage: conceder(c.statistics),
      personalization_storage: conceder(c.preferences),
      functionality_storage: 'granted',
      security_storage: 'granted',
    });
    w.dataLayer = w.dataLayer || [];
    ['preferences', 'statistics', 'marketing'].forEach(function (k) {
      if (c[k]) w.dataLayer.push({ event: 'cookie_consent_' + k });
    });
    w.dataLayer.push({ event: 'cookie_consent_update' });
  }

  // -------------------------------------------------------------- interfaz

  var panel = d.getElementById('cqx-consent-panel');
  var casillas = {
    preferences: d.getElementById('cqx-c-preferences'),
    statistics: d.getElementById('cqx-c-statistics'),
    marketing: d.getElementById('cqx-c-marketing'),
  };
  var ultimoFoco = null;
  var icono = d.getElementById('cqx-consent-icono');

  /* El icono queda flotando en cuanto se cierra el modal, como el de Cookiebot.
     No es decorativo: es la única forma de volver a abrirlo y de retirar el
     permiso, y retirarlo tiene que ser tan fácil como darlo. Solo aparece donde
     se ha preguntado; donde no se pregunta no hay nada que reconsiderar. */
  function pintarIcono() {
    if (!icono) return;
    icono.hidden = !(cfg.aplica && caja.hidden);
  }

  function mostrar() {
    ultimoFoco = d.activeElement;
    caja.hidden = false;
    pintarIcono();
    var primero = caja.querySelector('button');
    if (primero) primero.focus();
  }

  function ocultar() {
    caja.hidden = true;
    pintarIcono();
    // Al cerrar, el foco va al icono: es donde acaba de quedar la acción, y
    // devolverlo al fondo de la página dejaría a quien navega con teclado sin
    // saber dónde está.
    if (icono && !icono.hidden) icono.focus();
    else if (ultimoFoco && ultimoFoco.focus) ultimoFoco.focus();
  }

  function decidir(c) {
    guardar(c);
    comunicar(c);
    ocultar();
  }

  function todas(valor) {
    return { preferences: valor, statistics: valor, marketing: valor };
  }

  function delPanel() {
    return {
      preferences: !!(casillas.preferences && casillas.preferences.checked),
      statistics: !!(casillas.statistics && casillas.statistics.checked),
      marketing: !!(casillas.marketing && casillas.marketing.checked),
    };
  }

  function pulsar(id, fn) {
    var b = d.getElementById(id);
    if (b) b.addEventListener('click', fn);
  }

  pulsar('cqx-aceptar', function () { decidir(todas(true)); });
  pulsar('cqx-rechazar', function () { decidir(todas(false)); });
  pulsar('cqx-personalizar', function () {
    var abierto = panel.hidden === false;
    panel.hidden = abierto;
    d.getElementById('cqx-personalizar').setAttribute('aria-expanded', String(!abierto));
  });
  pulsar('cqx-guardar', function () { decidir(delPanel()); });
  pulsar('cqx-consent-icono', function () { w.cqxConsent.abrir(); });

  // Sin botón de cerrar: cerrarlo sin elegir equivaldría a un consentimiento
  // que nadie ha dado. Escape tampoco lo cierra, por lo mismo.

  // Para el enlace de "configurar cookies" del pie: vuelve a abrirlo con lo que
  // haya guardado ya marcado, que es lo que exige poder retirar el permiso.
  w.cqxConsent = {
    abrir: function () {
      var y = leer();
      Object.keys(casillas).forEach(function (k) {
        if (casillas[k]) casillas[k].checked = y ? !!y[k[0]] : false;
      });
      panel.hidden = false;
      d.getElementById('cqx-personalizar').setAttribute('aria-expanded', 'true');
      mostrar();
    },
    estado: leer,
  };

  // ---------------------------------------------------------------- arranque

  var guardado = cfg.forzar ? null : leer();   // con ?debug=1 se ignora lo guardado
  if (guardado) {
    comunicar({ preferences: !!guardado.p, statistics: !!guardado.s, marketing: !!guardado.m });
    pintarIcono();
    return;
  }
  if (!cfg.aplica) {
    // Fuera de las regiones con normativa de consentimiento previo no se
    // pregunta y se da por implícito, que es lo que hace Cookiebot hoy
    // (`method: "implied"`). No se guarda cookie: no hay decisión que guardar.
    comunicar(todas(true));
    return;
  }
  mostrar();
})(window, document);

File: calendario/funnels/consentimiento.py
# -*- coding: utf-8 -*-
"""Consentimiento de cookies propio, en sustitución de Cookiebot.

Cookiebot lo inyecta el contenedor de GTM de cada marca, no este código, así que
en las páginas que sirve Django convivían dos cosas: su banner —con un diseño
que no encaja con ninguna de las tres marcas— y todo lo demás. Aquí se monta el
nuestro y se impide que el suyo se cargue en estas páginas.

Se replica su comportamiento, no solo su aspecto:

- Las mismas cuatro categorías: necesarias (siempre), preferencias,
  estadísticas y marketing.
- Dos modos, según desde dónde se entre, porque la ley no es la misma:

  · EXPLÍCITO en el Espacio Económico Europeo, Reino Unido, Suiza y Brasil. El
    RGPD exige consentimiento previo e inequívoco: hasta que no se pulsa, no se
    activa nada. Es lo que hace el modo por defecto de Consent Mode.

  · IMPLÍCITO en el resto: LATAM (salvo Brasil) y Estados Unidos. Ahí el aviso
    también se muestra, pero no bloquea: informa de que «al continuar
    navegando, aceptas su uso», el consentimiento se concede desde el principio
    y se cierra solo al seguir navegando. California y los estados con leyes
    parecidas son de EXCLUSIÓN, así que lo que exigen no es permiso previo sino
    poder retirarlo, y para eso está el botón de configurar y el icono.

  El aviso sale SIEMPRE; lo que cambia es el modelo. Si se comprueba contra
  Cookiebot y en alguna región no aparece nada, mirar su configuración antes de
  tocar esto: tenía los países de LATAM desmarcados, y por eso no salía.
- Google Consent Mode v2, con los mismos eventos al dataLayer que empuja su
  plantilla de GTM (`cookie_consent_<categoría>` por cada una aceptada y
  `cookie_consent_update` al final), para que los triggers que ya existen en los
  contenedores sigan funcionando sin tocarlos.

Lo que se pierde respecto a Cookiebot y conviene tener presente: su registro de
consentimientos como prueba ante una inspección, y el escaneo automático que
mantiene al día la tabla de cookies de la política de privacidad. Lo primero se
cubre guardando la decisión con fecha y versión; lo segundo pasa a ser trabajo
manual.
"""

# Países donde se pregunta antes de activar nada.
#
# Los 27 del EEE más Islandia, Liechtenstein y Noruega (RGPD por el acuerdo del
# Espacio Económico Europeo), Reino Unido (UK GDPR), Suiza (nLPD) y Brasil
# (LGPD, que también es de consentimiento previo).
#
# NO cubre California: la CCPA es de exclusión —basta un enlace de "no vendas
# mis datos", no un banner previo— y además `CF-IPCountry` da país, no estado,
# así que gatearla aquí significaría preguntar a todo Estados Unidos.
PAISES_CON_CONSENTIMIENTO_PREVIO = frozenset({
    # Unión Europea
    'AT', 'BE', 'BG', 'CY', 'CZ', 'DE', 'DK', 'EE', 'ES', 'FI', 'FR', 'GR',
    'HR', 'HU', 'IE', 'IT', 'LT', 'LU', 'LV', 'MT', 'NL', 'PL', 'PT', 'RO',
    'SE', 'SI', 'SK',
    # Resto del Espacio Económico Europeo
    'IS', 'LI', 'NO',
    # Reino Unido y Suiza
    'GB', 'CH',
    # Brasil
    'BR',
})

# Sube cuando cambien las categorías o los textos: un consentimiento guardado
# con una versión anterior se vuelve a pedir, porque ya no se dio sobre lo mismo.
VERSION = 1

# Paleta por marca. El banner es el mismo componente en las tres, pero cada una
# tiene su propio lenguaje visual y hay que hablarlo, o vuelve a parecer —como
# el de Cookiebot— una pieza pegada de otro sitio:
#
#   papel → tarjeta de cartón con su textura, como el resto de tarjetas de
#           Blocks y Finance
#   pixel → CTA con el borde pixelado y el degradado de marca, el mismo del
#           «Ver vídeo gratis» del funnel
#
# Languages no usa ninguna de las dos: sus tarjetas son blancas y sus botones
# píldoras verdes, así que ahí el componente va liso y redondeado.
MARCAS = {
    'conquer-blocks': {
        'politica_url': 'https://www.conquerblocks.com/legal/politica-de-privacidad',
        'acento': '#ff4000',
        'acento_texto': '#ffffff',
        'fuente': "'Funnel Display',Arial,sans-serif",
        'radio': '10px',
        'papel': True,
        'pixel': True,
        'grad_1': '#ff4000',
        'grad_2': '#ff9800',
    },
    'conquer-finance': {
        'politica_url': 'https://www.conquerfinance.com/legal/politica-de-privacidad',
        'acento': '#3ac043',
        'acento_texto': '#ffffff',
        'fuente': "'Funnel Display',Arial,sans-serif",
        'radio': '10px',
        'papel': True,
        'pixel': True,
        'grad_1': '#aed916',
        'grad_2': '#3ac043',
    },
    'conquer-languages': {
        'politica_url': 'https://www.conquerlanguages.com/politica-de-privacidad',
        'acento': '#15b961',
        'acento_texto': '#ffffff',
        'fuente': 'Poppins,Arial,sans-serif',
        'radio': '20px',
    },
    'conquer-legal': {
        # Legal cuelga sus textos legales de /legal/; sin ese tramo la URL da
        # 404, no una redirección como en Blocks y Finance.
        'politica_url': 'https://www.conquerlegal.com/legal/politica-de-privacidad',
        'acento': '#0040FF',
        'acento_texto': '#ffffff',
        'fuente': "'Funnel Display',Arial,sans-serif",
        'radio': '10px',
        'papel': True,
        'pixel': True,
        # El CTA de Legal no es un degradado de dos paradas a 135deg como el de
        # Blocks: son tres, en horizontal, de periwinkle a navy. Se copia entero
        # en vez de aproximarlo con `grad_1`/`grad_2`.
        'gradiente': 'linear-gradient(90deg,#3E76FF 0%,#1845D6 42%,#031464 100%)',
    },
    # La corporativa del grupo. No es una escuela y no comparte su lenguaje:
    # fondo blanco liso, sin la textura de cartón, y el CTA es un rectángulo
    # gris oscuro casi sin redondear — nada de degradados ni de esquinas
    # pixeladas. Valores sacados del botón «Contacto» de www.conquerx.com:
    # `background rgb(51,51,51)`, `color #fff`, `border-radius 2px`, sin borde
    # y sin `clip-path`. Como los botones del banner se redondean con
    # `calc(var(--radio) - 2px)`, el radio va a 4px para que salgan a 2.
    'conquerx': {
        'politica_url': 'https://www.conquerx.com/politica-de-privacidad',
        'acento': '#333333',
        'acento_texto': '#ffffff',
        'fuente': "'Funnel Display',Arial,sans-serif",
        'radio': '4px',
    },
}

_POR_DEFECTO = {
    'politica_url': '/politica-de-privacidad',
    'acento': '#171717',
    'acento_texto': '#ffffff',
    'fuente': 'system-ui,-apple-system,Segoe UI,Roboto,Arial,sans-serif',
    'radio': '10px',
}


def modo(request):
    """`explicito` donde hace falta permiso previo; `implicito` en el resto.

    Se puede forzar con `?consent=eu` o `?consent=row` para poder revisar los
    dos sin fingir una IP, que es la única forma que había de verlos.
    """
    forzado_modo = (request.GET.get('consent') or '').lower()
    if forzado_modo in ('eu', 'explicito'):
        return 'explicito'
    if forzado_modo in ('row', 'implicito'):
        return 'implicito'
    return 'explicito' if _pide_permiso_previo(request) else 'implicito'


def forzado(request):
    """`?debug=1` saca el banner aunque no toque.

    Desde LATAM o US no se muestra —y ahí trabajamos casi siempre—, así que sin
    esto no hay forma de verlo ni de repasar cómo queda en cada marca sin
    fingir una IP europea. Además ignora la decisión ya guardada: si no,
    aceptaría una vez y no volvería a salir.
    """
    return request.GET.get('debug') == '1'


# Cabecera por la que llega el país del visitante.
#
# NO es `CF-IPCountry`. Cloudflare la reserva: el Worker que enruta los dominios
# de marca hacia Django corre en su red, y ahí los nombres que empiezan por
# `CF-` los gestiona Cloudflare —si el Worker intenta fijarla en la subpetición,
# se descarta—. Tampoco la pone Cloudflare por su cuenta, porque el origen
# (`calendar.conquerx.com`) no está detrás de Cloudflare: la añade al proxear
# hacia orígenes de sus propias zonas, y esta no lo es.
#
# Así que el Worker la reenvía con un nombre propio:
#
#     headers.set("X-Visitor-Country", request.cf?.country || "XX");
#
# Se sigue aceptando `CF-IPCountry` por si algún día el origen pasa a estar
# detrás de Cloudflare y la pone él.
CABECERA_PAIS = 'X-Visitor-Country'


def pais(request):
    """Código ISO del país del visitante, o cadena vacía si no se sabe."""
    return (request.headers.get(CABECERA_PAIS)
            or request.headers.get('CF-IPCountry')
            or '').upper()


def _pide_permiso_previo(request):
    """¿Está mirando desde un sitio donde hay que pedir permiso antes?

    Si no llega el país —fuera de Cloudflare, en local, o con la cabecera
    perdida por el camino— se asume que sí: ante la duda, pedir permiso es lo
    correcto y lo barato. El precio de equivocarse es enseñar el aviso
    bloqueante a quien no le tocaba; al revés sería medir sin permiso.
    """
    codigo = pais(request)
    if not codigo or codigo in ('XX', 'T1', 'T2'):
        return True
    return codigo in PAISES_CON_CONSENTIMIENTO_PREVIO


def aplica(request):
    """Siempre se enseña algo. Lo que cambia con la región es el modelo."""
    return True


def marca(escuela):
    return MARCAS.get((escuela or '').strip().lower(), _POR_DEFECTO)


def contexto(request, escuela=None):
    m = marca(escuela)
    return {
        'aplica': aplica(request),
        'modo': modo(request),
        'explicito': modo(request) == 'explicito',
        'forzado': forzado(request),
        'version': VERSION,
        'marca': m,
        'politica_url': m['politica_url'],
    }
